### DISCUSSION TITLE
Environment variables behaving differently within Quarkus extensions?

### DISCUSSION THREAD

**Discussion by Tomnivore** [quarkusio/quarkus#44663] [Q&A] [answered]
It seems there's a disparency with how different extension in Quarkus read ENV file variables.
My examples are examples are quarkus-rest, quarkus-rest-client-oidc-filter and quarkus-logging-gelf.

If I do a service with `@RegisterRestClient("my-client")` `QUARKUS_REST_CLIENT_MY_CLIENT_URL` works directly (as long as there are no capital letters). This is how I expect it to work

However if put a `@OidcClientFilter("my-oauth")` `QUARKUS_OIDC_CLIENT_MY_OAUTH_CLIENT_ID` does not connect to the config without also declaring it in `application.properties` as `quarkus.oidc-client.my-oauth.client-id`.

Likewise `QUARKUS_LOG_HANDLER_GELF_ADDITIONAL_FIELD_CUSTOM_FIELD_VALUE` won't work without declaring `quarkus.log.handler.gelf.additional-field.custom-field.value`.

Is there a reason or is it an oversight? If it's an oversight is it something that could and would be fixed?

**Comment by radcortez** [comment, marked as answer]
When you have something like `QUARKUS_REST_CLIENT_MY_CLIENT_URL`, where `MY_CLIENT` is a dynamic name, we cannot reliably determine only with this information if `MY_CLIENT` should be `my-client` (dash), `my.client` (dot), or even `my_client` (underscore): https://quarkus.io/guides/config-reference#environment-variables

Now, the REST Client config did work without having to declare the dotted version because it was just querying all possible combinations, which is not a desired behavior. See:
- https://github.com/quarkusio/quarkus/pull/42106
- https://github.com/quarkusio/quarkus/pull/42932

Because we didn't want to remove that behaviour, we added some proper support for it with:
- https://github.com/smallrye/smallrye-config/pull/1220
- https://github.com/quarkusio/quarkus/pull/43345

Such extensions like `quarkus-rest-client-oidc-filter` or `quarks-logging-gelf` must discover the names during build time and feed them to the config so we know which keys to look for. If not, the configuration would still need additional help by adding the exact name to a non-environment variable source to match it with its environment variable name.

**Reply by Tomnivore** [reply]
Thank you for the answer. Although it could be cool to be able to do all the stuff with ENVs, I understand your answer and it's reasonable. I will mark this question answered.
